Fixes #32605 - normalize subnet address

lzap opened the pull request:
User who accidentally entered incorrect subnet address was struggling with DHCP conflicts, because subnet address must match exactly with what's set in DHCP which obviously does not allow incorrectly formatted subnet addresses.

This ensures that subnet address is masked before saving, so there are no trail nonsense numbers. It ignores incorrectly entered addresses as they are handled in different validators.

https://community.theforeman.org/t/deleting-a-provisionned-host-then-rediscover-and-reprovision-fails-at-dhcp/23422/29

<!---

Thank you for contributing to The Foreman, please read the
[following guide](https://www.theforeman.org/contribute.html), in short:

* [Create an issue](https://projects.theforeman.org/projects/foreman/issues)
* Reference the issue via `Fixes #1234` in the commit message
* Prefer present-tense, imperative-style commit messages
* Mark all strings for translation, see [1]
* Suggest prerequisites for testing and testing scenarios following example above.
* Prepend `[WIP]` for work in progress to prevent bots from triggering actions
* Be patient, we will do our best to take a look as soon as we can
* Explain the purpose of the PR, attach screenshots if applicable
* List all prerequisites for testing (e.g. VMware cluster, two smart proxies...)
* Reviewers often use extensive list of items to check, have a look prior submitting [2]
* Be nice and respectful

1: https://projects.theforeman.org/projects/foreman/wiki/Translating#Translating-for-developers
2: https://github.com/theforeman/foreman/blob/develop/developer_docs/pr_review.asciidoc
-->

lzap committed:
Fixes #32605 - normalize subnet address

evgeni commented on test/models/subnet_test.rb:
```diff
@@ -102,13 +102,13 @@ class Subnet::Test < Subnet; end
-    assert_equal "10.0.0.22", s.network
+    assert_equal "10.0.22.0", s.network
     assert_equal "255.255.255.0", s.mask
     assert_equal "10.0.0.138", s.gateway
```
this gateway is not in 10.0.2.0/24, shouldn't it also be rejected? :)

lzap replied:
While I agree, I would not like to go crazy with this one. At least we should have ISC DHCP error about this one, not just REST API 404 that confuses Foreman.

evgeni replied:
Yeah, but at least make the test use sensible data ;)

lzap replied:
Yeah it is easy to do, let me add this validation then.

ekohl replied:
I don' think we need validation tbh. There's some edge cases like point to point connections. From my production `/etc/network/interfaces`:
```
auto eth0
iface eth0 inet static
        address 94.142.242.30/32
        gateway 94.142.244.11
        pointtopoint 94.142.244.11
```

This results in:
```
$ ip a show dev eth0
2: eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc pfifo_fast state UP group default qlen 1000                                                         
    link/ether 52:54:00:09:52:90 brd ff:ff:ff:ff:ff:ff
    inet 94.142.242.30/32 brd 94.142.242.30 scope global eth0
       valid_lft forever preferred_lft forever
    inet6 2a02:898:210::30/64 scope global
       valid_lft forever preferred_lft forever
    inet6 fe80::5054:ff:fe09:5290/64 scope link
       valid_lft forever preferred_lft forever
$ ip r
default via 94.142.244.11 dev eth0 onlink
```

evgeni replied:
Brrrr. But yeah, that's valid (and a crude way to save IP addresses).

I really just meant to update the test to use a `10.0.22.0/24` gateway, that's all.

ekohl replied:
:+1: to that

lzap replied:
Yeah missed that.

ekohl commented on lib/net/validations.rb:
```diff
@@ -119,6 +119,12 @@ def self.normalize_ip6(ip)
 
+    def self.normalize_network_address(ip, mask)
+      IPAddr.new(ip).mask(mask).to_s
+    rescue
```
Should this log something? Probably at the info level.

lzap replied:
Done.

evgeni commented on lib/net/validations.rb:
```diff
@@ -119,6 +119,12 @@ def self.normalize_ip6(ip)
     end
 
+    def self.normalize_network_address(ip, mask)
+      IPAddr.new(ip).mask(mask).to_s
```
```suggestion
      IPAddr.new("#{ip}/#{mask}").to_s
```
Saves creating one IPAddr instance by directly using the right mask.

evgeni replied:
`to_s` compresses v6 addresses:

```console
irb(main):007:0> IPAddr.new('2600::0').to_s
=> "2600::"
```

not sure this is intended here.

ekohl replied:
I do think it's good to compress. Makes it easier to read.

evgeni replied:
While I technically agree (and would always use the compressed notation), I also kinda think we shouldn't override the users preference?

ekohl replied:
That did make me think about where to solve it. In https://github.com/theforeman/foreman/pull/8528#issuecomment-842221068 I made the argument that we can also normalize it when using the value.

On the other hand, now you can have multiple Subnet instances represent the same subnet (`2600::0` and `2600::`) without the DB knowing. Now that we're PostgreSQL only, we can investigate moving to the native IP representation so that they're also properly sorted etc, but that's probably a bigger task. I think that the DB model layer also returns an IPAddr instance then which the Foreman code may not handle right now.

lzap replied:
I am not replacing two fast method calls with an interpolation. That is actually slower and more error prone :-)

Anyway, valid point to the compression. While I lean towards doing it, in case you don't like it then I can turn this into validation so an error is thrown if there are any bits in the non-mask section of the address.

ekohl commented:
Another thought: this won't fix up any existing DB entries. We should probably also make the outgoing request to Foreman Proxy validated. For example here:
https://github.com/theforeman/foreman/blob/e49da5764920b3810e1c33ea2517ec8c870362b8/lib/proxy_api/dhcp.rb#L36

Should we introduce a property on `Subnet` that's an instance of `IPAddr` that can be used elsewhere? Like `network_addr`

ekohl reviewed (changes requested):
Making it clear this is waiting on @lzap now.

lzap commented:
Added the recommended helper method so Proxy API uses masked subnet address to do DHCP calls.. Rebased. Still there is the open question of compression of IPv6 addresses.

ekohl commented on app/models/subnet.rb:
```diff
@@ -396,6 +402,20 @@ def ensure_ip_addrs_valid
+    network
+  end
+
+  def ensure_gateway_in_subnet
```
As noted in https://github.com/theforeman/foreman/pull/8528#discussion_r636880287 I think this is wrong. I would need a way to disable this validation.

lzap commented:
Fixed tests hopefully.

evgeni commented on app/models/subnet.rb:
```diff
@@ -396,6 +409,13 @@ def ensure_ip_addrs_valid
   end
 
+  def ensure_gateway_in_subnet
+    masked_gw = IPAddr.new(gateway).mask(mask)
+    errors.add(:gateway, _("is not from the subnet")) unless masked_subnet_address == masked_gw
```
why not:

```suggestion
    gw = IPAddr.new(gateway)
    errors.add(:gateway, _("is not from the subnet")) unless masked_subnet_address.include?(gw)
```

but then see above… (masked subnet is not always an `IPAddr`)

lzap replied:
Oh yeah your solution is definitely better than mine. Not sure what you mean "see above". Oh perhaps the below comment, we should definitely return the same type. I will fix this.

evgeni replied:
Yeah, I meant the type comment, thanks GH for sorting it differently ;)

evgeni commented on app/models/subnet.rb:
```diff
@@ -363,6 +365,13 @@ def dns_servers
   end
 
+  def masked_subnet_address
+    IPAddr.new(network).mask(mask)
+  rescue StandardError => e
+    logger.info "Network address #{network} cannot be masked: #{e}"
+    network
```
this returns an `IPAddr` in the case it could be masked, and a `String` otherwise, which feels odd?

evgeni replied:
`Net::Validations.normalize_network_address` always returns a string and is otherwise identical to this function?

lzap commented:
Okay hopefully fixed all the tests and amended both your remarks, thanks!

evgeni commented on lib/net/validations.rb:
```diff
@@ -119,6 +119,13 @@ def self.normalize_ip6(ip)
+    def self.normalize_network_address(ip, mask)
+      IPAddr.new(ip).mask(mask).to_s
+    rescue StandardError => e
+      Foreman::Logging.logger('app').info "Unable to validate gateway address: #{e}"
```
*gateway*? this method is about subnets.

lzap replied:
C&P error, fixing.

ekohl commented on app/models/subnet.rb:
```diff
@@ -102,8 +102,10 @@ def model_name
+  before_validation :normalize_subnet
   after_validation :validate_against_external_ipam
   validate :ensure_ip_addrs_valid
+  validate :ensure_gateway_in_subnet
```
I don't think this should be in this PR. As I noted: it makes it more strict and isn't always true. Granted, it's an edge case but it at least deserves its own entry in the changelog and probably an upgrade warning. To get this PR merged, it's best left out of it.

lzap replied:
Yeah I misunderstood Evgeni above, you are right. Removed.

evgeni commented on app/models/subnet.rb:
```diff
@@ -396,6 +409,13 @@ def ensure_ip_addrs_valid
 
+  def ensure_gateway_in_subnet
+    gw = IPAddr.new(gateway)
+    errors.add(:gateway, _("is not from the subnet")) unless masked_subnet_address.include?(gw)
```
this makes "out of subnet" gateways as @ekohl showed in his example impossible.

I'm personally fine with being that strict, but Ewoud might think differently :)

ekohl replied:
As noted elsewhere, I certainly don't think it belongs in this PR because users may be surprised. As its own PR, we can discuss the impact.

lzap replied:
Yup, I did not think it through. We do not totally enforce that, makes no sense.

lzap commented:
Rebased, amended the typo and removed enforcement of gateway from subnet nonsense. Hopefully we are green now.

evgeni commented on lib/net/validations.rb:
```diff
@@ -119,6 +119,13 @@ def self.normalize_ip6(ip)
+    def self.normalize_network_address(ip, mask)
+      IPAddr.new(ip).mask(mask).to_s
+    rescue StandardError => e
+      Foreman::Logging.logger('app').info "Unable to validate network address: #{e}"
```
The method is called normalize, not validate, so the message seems wrong?

Also I wonder under which circumstances we can get an error and yet still want to continue using that address?

lzap replied:
The method needs to be called via `before_validate` so it is normalized early in the callback so orchestration already has it in the correct format. There are two options:

1) Raise an exception which will render to the ugly 500 page,
2) Log an error and return something. That is the path I have chosen, I do believe there will only be exception when something got utterly wrong, for example netmask does not make sense or the input is not a valid IP address which will be caught later on via `ensure_ip_addrs_valid`

evgeni commented on app/models/subnet.rb:
```diff
@@ -363,6 +364,13 @@ def dns_servers
+  def masked_subnet_address
+    IPAddr.new(network).mask(mask)
+  rescue StandardError => e
+    logger.info "Network address #{network} cannot be masked: #{e}"
```
Same q as in the other hunk: what do we need to do to make mask() fail (and IPAddr.new succeed!) but still use that address?

lzap commented:
I have amended the message reformatting.

evgeni commented:
```
<p><span style="font-weight:bold">ComputeOrchestrationTest.test_0001_if MAC is changed, dhcp_record cache is dropped</span></p><div style="text-align:right;">Failing for the past 1 build
          (Since <a href="https://ci.theforeman.org/job/test_develop_pr_core/8457/database=postgresql,ruby=2.5,slave=fast/" class="model-link"><span style="width: 16px; height: 16px; " class="build-status-icon__wrapper icon-red icon-sm"><span class="build-status-icon__outer"><svg viewBox="0 0 24 24" focusable="false" class="svg-icon "><use href="https://ci.theforeman.org/images/build-status/build-status-sprite.svg#build-status-static"></use></svg></span><svg viewBox="0 0 24 24" focusable="false" class="svg-icon icon-red icon-sm"><use href="https://ci.theforeman.org/static/fe3dff5d/images/build-status/build-status-sprite.svg#last-failed"></use></svg></span>#8457</a> )
        </div><div style="text-align:right;"><a href="https://ci.theforeman.org/job/test_develop_pr_core/8457/database=postgresql,ruby=2.5,slave=fast/testReport/junit/(root)/ComputeOrchestrationTest/test_0001_if_MAC_is_changed__dhcp_record_cache_is_dropped/history">Took 0.11 sec.</a></div><div id="description"><div></div><div align="right"><a id="description-link" href="https://ci.theforeman.org/job/test_develop_pr_core/8457/database=postgresql,ruby=2.5,slave=fast/testReport/junit/(root)/ComputeOrchestrationTest/test_0001_if_MAC_is_changed__dhcp_record_cache_is_dropped/editDescription" onclick="return replaceDescription();"><img src="https://ci.theforeman.org/static/fe3dff5d/images/16x16/notepad.png" style="width: 16px; height: 16px; " class="icon-notepad icon-sm">add description</a></div></div>
<h3>Error Message</h3><pre>NameError: undefined local variable or method `ip' for #&lt;Subnet::Ipv4:0x0000000013de90c8&gt;...</pre><h3>Stacktrace</h3><pre>NameError: undefined local variable or method `ip' for #&lt;Subnet::Ipv4:0x0000000013de90c8&gt;
    app/models/subnet.rb:370:in `rescue in masked_subnet_address'
    app/models/subnet.rb:367:in `masked_subnet_address'
    app/models/concerns/orchestration/dhcp.rb:123:in `dhcp_attrs'
    app/models/concerns/orchestration/dhcp.rb:105:in `build_dhcp_record'
    app/models/concerns/orchestration/dhcp.rb:30:in `block in dhcp_records'
    app/models/concerns/orchestration/dhcp.rb:29:in `map'
    app/models/concerns/orchestration/dhcp.rb:29:in `dhcp_records'
    test/models/orchestration/compute_test.rb:109:in `block in &lt;class:ComputeOrchestrationTest&gt;' (NameError)
/usr/local/rvm/gems/ruby-2.5.1@test_develop_pr_core-1/gems/activemodel-6.0.3.7/lib/active_model/attribute_methods.rb:432    </pre>
```

looks related?

ekohl commented on test/models/subnet_test.rb:
```diff
@@ -102,15 +102,16 @@ class Subnet::Test < Subnet; end
     s.dns_secondary = " 10.0.0.60   "
-    assert s.save
-    assert_equal "10.0.0.22", s.network
+    s.save
```
Why should the assert be dropped here?

lzap replied:
Probably a leftover, putting back.

ekohl commented on test/models/subnet_test.rb:
```diff
@@ -102,15 +102,16 @@ class Subnet::Test < Subnet; end
   test "should strip whitespace before save" do
     s = subnets(:one)
-    s.network = " 10.0.0.22   "
+    s.network = " 10.0.22.0   "
```
Right now we have no test to normalize the network. You can also keep the changes smaller and keep using `network = " 10.0.0.22  "` and assert that it uses `10.0.0.0` as the network later on.

lzap replied:
Right, let me create an explicit test case.

ekohl commented on app/models/concerns/orchestration/dhcp.rb:
```diff
@@ -120,7 +120,7 @@ def dhcp_attrs(record_mac)
       :mac => record_mac,
       :proxy => subnet.dhcp_proxy,
-      :network => subnet.network,
+      :network => subnet.masked_subnet_address,
```
This changes the data type from `String` to `IPAddr`. Is that handled correctly elsewhere or should it use a string here?
```suggestion
      :network => subnet.masked_subnet_address.to_s,
```

lzap replied:
Good point.

lzap commented:
Thanks, fixed tests hopefully.

ekohl commented on app/models/subnet.rb:
```diff
@@ -363,6 +364,13 @@ def dns_servers
+  def masked_subnet_address
+    IPAddr.new(network).mask(mask)
+  rescue StandardError => e
+    logger.info "Unable to mask '#{network}' with '#{mask}': #{e}"
+    IPAddr.new(network)
```
Overall I think this is ok. The only thing I'd wonder is if we have some way to validate the subnet so we don't hit this. `normalize_subnet` is called in `before_validation` so it can certainly be hit. I also don't see a validates check on subnet other than presence. Perhaps a future improvement? In that case we probably also want to validate all records in the DB so subnets are valid.

ekohl replied:
Any thoughts on this?

ekohl reviewed (commented):
I'm not sure if these are real failures or if auditing tests on UserGroup are broken. It feels unrelated.

lzap commented:
Let's try CI again not sure what was wrong.
